### ISSUE TITLE
Add session client command

### ISSUE THREAD

**Issue by timdrysdale** [practable/relay#32]
## What is proposed?
A new command `session client` that connects to `session relay`, offering functionality including

### Core features
- log data to file (or stdout, or stderr), and re-open file on SIGHUP to facilitate [logrotate(8)](https://linux.die.net/man/8/logrotate)
- read data from file and send it to `relay`, line-by-line, at predefined intervals listed in the file
- connect to a USB device, with given properties for baudrate, stop bit, parity, etc.
- read data from stdin and send to `relay`

### Optional extras

These are straightforward to implement, although it may be efficient to wait until there is demand before implementing

- provide an insecure websocket server at a given port, so that other codes can access `relay` without being aware of the `access `
- provide an insecure HTTP server which can report latest messages received in response to a GET command (don't clear buffer as GET is idempotent)
  - Use POST to get latest commands, and clear buffer (to avoid case of receiving data between last GET and the following POST to clear buffer)
  - Use POST to send data to the relay (and maybe wait for a response) 

### What is NOT proposed
A client connecting to multiple relay topics, using rules to set the connections and parameters such as file to log to. This seems nice in principle, but it confounds having different logrotate policies for topics with different data rates, because signals can only be sent to the main process, not individual topics. The HTTP API could of course be used for triggering per-rule logrotate response, but this would require the `cron` jobs to start curl to send the commands. So it would be doable, but it also adds significant overhead to the system [if using many curl commands](https://stackoverflow.com/questions/37731609/curl-dont-wait-for-response) although one answer suggests 'netcat' with '--send-only' 
It's tempting to think this approach might make logrotation easier on windows, but similar overhead considerations apply if you are logging from dozens of experiments.

## Why do we need this?

Issues #16  and #10  both imply the need to connect to session host. Currently, we can do this with [timdrysdale/check-experiment-script](https://github.com/timdrysdale/check-experiment-script), which connects using bash scripting (see below), however this doesn't fully solve our problems because:

- we cannot rely on bash being present either on the experiment host (due to choice or limited resources) or for an end-user (e.g. may want to connect from a windows machine)
- we're running into issues piping from websocat into a file, when it runs in a background process (hence the gymnastics in the scripts to keep websocat in the foreground process).

We already have a model to follow from `cmd/shell`, which has a `client` command. Implementing a client command is a useful step on the way to solving #16, #10, and helps out with logging data.

## Potential Issues

- USB library could be linux-specific, preventing compilation on windows?
  - Solution: pick cross-platform library such as [gousb](https://github.com/google/gousb)

- Multiple USB devices with the same VID, PID
  - Solution: [pick the first one found](https://pkg.go.dev/github.com/google/gousb) as per the complex example.
  - Note this will result in undefined behaviour if duplicate devices are present.
  - Defer handling duplicate devices to the response to #10 (which will be able to build on the work we do in this issue)

- Log rotation on windows might not operate in the same way, e.g. optimal choice for signalling depending logrotate method?
  - Solution: 
     - look for windows implementation of logrotate e.g. [logrotatewin](https://github.com/plecos/logrotatewin)  although note recent punresolved reports of issues](https://sourceforge.net/projects/logrotatewin/)
     - architecture specific code to handle this issue, e.g. handle logrotation internally or use different signal to trigger new file
     

## Implementation suggestions

This section proposes how to implement the commands

### Separate USB and file handling
It is not particularly likely that USB and file interfaces would be used in the same command because it conflates something that should last the lifetime of the system (USB connections), and something which is inherently much shorter and likely to be repeated multiple times (running a check on the USB system). 

### Don't offer secure listening
All listening is expected to be local because the relay is there to provide other remote parties with access. So if HTPPS or WSS is somehow required, then the service should be proxied. Note that requiring proxying like this in your lab infrastructure is probably an "architecture smell" indicating to you that you are creating additional management burdens. There may be cases where proprietary code is being used to analyse data and it is enforcing secure connections about of an abundance of caution, so making a secure local connection would not be a "smell" so much as a justifiable inconvenience.

### Environment variables
As for other commands, environment variables are a convenient way to set parameters for system utilities, especially when some parameters are long, like the access token. The only constraint this imposes is that multiple copies of client have be run such that their local environment variable is correct, and this is readily done if a 'systemd' service, or in a shell script, or even by passing specific commands as environment variables in the same command line

```bash
export SESSION_CLIENT_TOKEN=ey....(rest of JWT token)
export SESSION_CLIENT_SESSION=https://access.example.io/session/abc123
export SESSION_CLIENT_DEVELOPMENT=true #omit setting this ENV VAR in production
```

### Commands

```bash
session client usb
session client file
session client ws-listen
session client http-listen
```

#### USB

required environment variables, e.g. for [ch340](https://www.mpja.com/download/35227cpdata.pdf)
```bash
export SESSION_CLIENT_USB_VID=0x1A86
export SESSION_CLIENT_USB_PID=0x7523
```

#### File

needs one or more of the following 

```bash
export SESSION_CLIENT_FILE_LOG=/var/log/session/spin35.log #typically omit if logging to stdin/stderr
export SESSION_CLIENT_FILE_PLAY=/etc/practable/checks/spinv2-short.json #omit if just logging
export SESSION_CLIENT_FILE_STDIN=true #typically omit if logging to file/stderr
export SESSION_CLIENT_FILE_STDERR=true #typically omit if logging to file/stdin
```

#### WS-listen

Listens on default port of 8081 unless otherwise specified, e.g:
```bash
export SESSION_CLIENT_WSLISTEN=8080
```

#### HTTP-listen
Listens on default port of 8080 unless otherwise specified, e.g:
```bash
export SESSION_CLIENT_HTTPLISTEN=5000
```
 
## Bash script example
This is the current solution which we are aiming to replace:

`timdrysdale/check-experimentp-script/scripts/connect.sh`
```bash
#!/bin/bash

# usage: ./connect.sh <expt> <stream>
# tokens and access files are assumed to be in ../autogenerated
# with naming convention:
# access file: <stream>.access.<expt>
# token file: <stream>.token.<expt>
#
# an example commmand: ./connect.sh spin30 log 20
# this puts incoming data to spin30-log-<timestamp>.json for 20 seconds
#
# timothy.d.drysdale@gmail.com
# 17 Jan 2022

dir=../autogenerated

# parse arguments
expt=${1}  #e.g. spin30
stream=${2} #e.g. log
duration=${3} #in bash time (support only seconds, minutes, hours, days)
outdir=${4}


# retrieve access
access=$(cat $dir/$stream.access.$expt)

# retrieve token
token=$(cat $dir/$stream.token.$expt)

# generate output filename with timestamp
suffix="json"

if [[ "$stream" == "video" ]]
then
	suffix="ts"
fi

out=$outdir/$expt-$stream-$(date +%s).$suffix

response=$(curl -X POST -H "Authorization: ${token}" $access --silent)

websocket=$(echo $response | jq -r '.uri')

if [[ "$websocket" == "null" ]] 
then
	echo "Error connecting to:" ${access}
	echo $response
	echo "token:" $token
	exit 1
fi

format="binary"

if [[ "$stream" == "video" ]]
then
	format="binary"
fi

#set -x # DEBUG ON

# Things that seem to only get a single message ...
# websocat --$format $websocket - > $out &
#{ websocat --$format $websocket - > $out; } &

# convert duration into seconds
# https://stackoverflow.com/questions/50433578/how-to-convert-time-period-string-in-sleep-format-to-seconds-in-bash
t2s() {
   sed 's/d/*24*3600 +/g; s/h/*3600 +/g; s/m/*60 +/g; s/s/\+/g; s/+[ ]*$//g' <<< "$1" | bc
}

seconds=$(t2s $duration)

#Use timeout script to avoid websocat only writing a single message to file when backgrounded

if [ -n "${5}" ]
then
	#test sh-c on command line websocat - sh-c:"./play.sh ../files/step.json"

	./timeout.sh -t $seconds websocat --$format $websocket sh-c:"./play.sh ${5} ${6}" #do not pipe to file (it will be empty)
	
else
    ./timeout.sh -t $seconds websocat --$format $websocket - > $out
fi

#set +x #DEBUG OFF


```

**Comment by timdrysdale** [comment]
@dpreid  reminded me that we will usually want to add timestamps when logging to file. 

This is probably best as an option. 

There are choices to be made for the format of the timestamp, and for each line in the log. The optimal choice is RFC3339 in syslog format, e.g.

```bash
[2022-03-10T13:34:43.777983000Z] {"some":"msg"}
```

## Timestamp format

Optimal choice is probably [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339) since it is readily machine readable, and human readable, and closer to what linux logs typically use.

To aid reading of log files, let's specify some requirements 
 
  - an uppercase "T" character MUST be used to separate date and time
  - an uppercase "Z" character MUST be present in the absence of a numeric time zone offset.
  - We will always include fractional seconds to the resolution of nanoseconds, so as to avoid repeated timestamps

(the first two requirements are from [here](https://validator.w3.org/feed/docs/error/InvalidRFC3339Date.html))
## Format

The other options were:
  - linux epoch at GMT - not very human readable, but easy enough to process 
  - time since startup - not desirable because we will want to correlate to the real date-time for analysis and avoid confusion if the logging machine is restarted
  - some other variation on a human readable timestamp - sure, we could, but there are no known issues with ambiguity in RFC3339 whereas other formats may introduce problems with timezones etc

## File format

The format needs to provide an unambiguous separation between timestamp and message. For readability, without sacrificing machine readability/parsing, a syslog format is preferred.

### Syslog style
Note space after the closing `]` for readability 
```bash
[2022-03-10T13:34:43.777983000Z] {"some":"msg"}
```

### JSON

This requires escaping the double quotes in the json, making human readability suffer.
```json
{"t":"2022-03-10T13:34:43.777983000+00:00", "m":"{\"some\":\"msg\"}"}
```

**Comment by timdrysdale** [comment]
Addressed in #34
